### DISCUSSION TITLE
Strongly typed Meteor methods/subscribes and combined

### DISCUSSION THREAD

**Discussion by chneau** [meteor/meteor#13160] [Feature Requests]
I had a trial by myself at that:

https://github.com/chneau/meteor-template/tree/master/imports

the two files:
- typed-method.ts
- typed-subscribe.ts

I'm planning to give it a go again with 3.0-rc2 in the following days. 

I just want simple typescript wrapper around methods and subscriptions.

Do you use TypeScript? What do you think of the `Meteor.call()` not being typed? What do you think of the `Meteor.subscribe("...")` not being typed?

**Comment by JorgenVatle** [comment]
I'm perhaps a bit late to the discussion, but just wanted to pitch in.

If you haven't seen it already do check out the [`zodern:relay`](https://github.com/zodern/meteor-relay) package. It has the most user-friendly and clever solutions I've seen to deal with the issue.

## Adding types to `Meteor.call()` and `Meteor.subscribe()`

You can type Meteor's `call` and `subscribe` methods without needing to wrap or change any of their underlying functionality. It does require you to structure your app slightly differently from how you normally would when using Meteor. This would still be the case even with native support from Meteor unless some sort of compilation step is added to generate type definitions as Meteor is running. 

Using a compilation step is not without drawbacks as it does require Meteor to be running in order for you to get type hints for newly defined resources. So like as both you and Zodern have found, the most reliable approach is defining resources as modules.

Using the same structure like you would when defining a method or publication, define it as a module instead.
```ts
// links.methods.ts
export default {
  'links.create': (link: LinkDocument) {
    return LinksCollection.insert(link)
  }
}

// links.publications.ts
export default {
  'links.find'(query: Mongo.Query<LinkDocument>) {
    return LinksCollection.find(query);
  }
}
```

Doing it this way instead of registering, say, methods directly with `Meteor.methods({ ... })` has a few benefits in terms of typing. Since they're static, we can infer types from them, something which isn't possible when dynamically calling a method like `Meteor.methods()` or `Meteor.publish()`.

From here, we can combine all our methods and publications into one large index module just to make referencing them later a little easier.
```ts
// api/index.ts
export const Methods = {
  ...LinkMethods,
  ...UserMethods,
  ...etc
}

export const Publications = {
    ...LinkPublications,
    ...etc
}
```

Now we have both our publications and methods including their types in one place. Now it's just a matter of augmenting Meteor's type definitions to know about these.

```ts
// api/index.ts
declare module 'meteor/meteor' {
  interface DefinedMethods extends typeof Methods {}
  interface DefinedPublications extends typeof Publications {}
}
```

And we can register these now like you always would with Meteor.
```ts
Meteor.methods(Methods);

// Unfortunately Meteor doesn't provide a Meteor.publications() method yet.. 😅
Object.entries(Publications).forEach(([name, publication]) => Meteor.publish(name, publication));
```

Assuming Meteor has added the types you will see in the following section, that would be all you have to do. When typing `Meteor.call('li...` in your editor, you will get auto-completion for your method names and warnings when passing the wrong type like you would normally expect.

## Types that could be beneficial in Meteor's type definitions

Optimally Meteor would provide empty `DefinedMethods` and `DefinedPublications` interfaces as part of their core type definitions like they do for other types they expect you to extend, like the `UserServices` and `UserProfile` interfaces in `meteor/meteor`. 

This would make it a lot more user friendly to add strong types to your Meteor app, as it would make the following type augmentation redundant.

```ts
// meteor.d.ts
declare module 'meteor/meteor' {
  type MethodName = keyof DefinedMethods;
  type MethodArgs<TName extends MethodName> = Parameters<DefinedMethods[TName]>
  type MethodResponse<TName extends MethodName> = ReturnType<DefinedMethods[TName]>
  type PublicationName = keyof DefinedPublications;
  type PublicationArgs<TName extends PublicationName> = Parameters<DefinedPublications[TName]>

  namespace Meteor {
    function call<TName extends MethodName>(method: TName, ...params: MethodArgs<TName>): MethodResponse<TName>;
    function subscribe<TName extends PublicationName>(publication: TName, ...params: PublicationArgs<TName>): Meteor.SubscriptionHandle;
  }
}

```
## Schemas
You can and probably should go a little further of course and define your resource arguments as schemas, like in your example, instead of plain typed. but not validated types. Inferring types with schemas gets a little more involved. But I think that's probably more the responsibility of library authors rather than Meteor's.

I have a [library using this approach with Valibot for schema validation](https://github.com/JorgenVatle/meteor-type-validation) if it's of any value as a reference. I would probably still recommend opting for [`zodern:relay`](https://github.com/zodern/meteor-relay) if you're looking for a solid type-safe alternative to Meteor's publication and method handles.
